### ISSUE TITLE
AdaptiveLayer Evaluation

### ISSUE THREAD

**Issue by bobox2997** [UKPLab/sentence-transformers#3329]
In the evaluation sections of the docs i see that there is an option to test the performance while truncating the dimensions (i assume it is to test models trained with MatryoshkaLoss). Is there anything similar that can be used to test models while training with AdaptiveLayerLoss? 
someone has any suggestions/advice about that?

thanks in advance!

**Comment by tomaarsen** [comment]
Hello!

I considered adding this support natively when I added the AdaptiveLayer loss, but it ended up not being as simple as I would have imagined. When Sentence Transformers loads a transformers-based model, it considers it a bit of a black box. This is convenient, as a lot of different architectures supported in `transformers` work out of the box in Sentence Transformers, but it also means that I don't always know what architecture is being used.

And I need to know which architecture is being used to be able to adaptively shrink the number of layers in the underlying model. With other words, the only way I can reasonably add this, is by manually implementing the "path" to the layers of the underlying model for every single `transformers` architecture. I wanted to avoid this, so I didn't implement it in Sentence Transformers.

If you're doing your own evaluation with a model trained with AdaptiveLayerLoss, then I would subclass an existing evaluator, and override the `__call__` method with something like this:
```python

class AdaptiveLayerEmbeddingSimilarityEvaluator(EmbeddingSimilarityEvaluator):
    def __init__(self, num_layers: int, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.num_layers = num_layers

    def __call__(self, model, *args, **kwargs):
        all_model_layers = model[0].auto_model.bert.layers # or whatever path you have to the model layers
        model[0].auto_model.bert.layers = all_model_layers[:self.num_layers]
        output = super().__call__(model, *args, **kwargs)
        model[0].auto_model.bert.layers = all_model_layers
        return output
```

and then you'll lower the number of layers for your evaluation.

- Tom Aarsen

**Comment by bobox2997** [comment]
hi @tomaarsen 

thank you for the reply, now I understand the reason!
I ended up subclassing an evaluator as you suggested and it work fine, thanks!



as i'm here..., I've been experimenting quite a bit with some slightly modified AdaptiveLayerLoss, and wanted to share a couple of observations from my training runs, just to see if they resonate or spark any discussion:

- I've sometimes found that the `1 / (1 + layer_idx)` factor applied to the `layer_loss` might occasionally impact training stability and slightly hinder the performance of final layers (more noticeably with higher `kl_div_weight`). I undertand that this is not always true, still, have you experimented making this weighting optional? (even reverting it may have some logic behind...)

- about "layer selection"...
    *   excluding the first N layers (e.g., the initial 10-20% of the model) he prior loss calculation seems to improve the performance of the subsequent layers being trained. it feels like keeping some layers 'free' is beneficial. 
    *   I've also seen some potential benefits from training on a non-continuous set of layers (e.g., every other layer, or a specific list). while the goal is adaptability, maybe offering specific, optimized 'cut-off' points via targeted training could be a useful tradeoff?

Just sharing these thoughts based on my experiences.

No pressure to act on these, just wanted to share what I was seeing in case it's useful feedback or aligns with others' experiences. Happy to discuss more if it makes sense... i'm just urious if this aligns with anything you've seen or considered. If these ideas seem potentially useful, I'd be happy to open a more formal feature request later.

As always, a big thanks to you and to the whole community!

**Comment by tomaarsen** [comment]
Apologies for the delay!

> * I've sometimes found that the `1 / (1 + layer_idx)` factor applied to the `layer_loss` might occasionally impact training stability and slightly hinder the performance of final layers. I undertand that this is not always true, still, have you experimented making this weighting optional? (even reverting it may have some logic behind...)

Hmm, looking at the code again now, I have a feeling like this:
```python
            loss = loss + layer_loss / (1 + layer_idx) / len(layer_indices) * self.prior_layers_weight
```
was actually intended as
```python
            loss = loss + layer_loss / ((1 + layer_idx) / len(layer_indices)) * self.prior_layers_weight
```
i.e. `((1 + layer_idx) / len(layer_indices))` to give extra weight to the later layers, but the lack of brackets actually seems like it does the opposite.

>* excluding the first N layers (e.g., the initial 10-20% of the model) he prior loss calculation seems to improve the performance of the subsequent layers being trained. it feels like keeping some layers 'free' is beneficial. 

Very fair! A lot of traditional finetuning involves freezing the first X layers and only finetuning the later ones. 


> * I've also seen some potential benefits from training on a non-continuous set of layers (e.g., every other layer, or a specific list). while the goal is adaptability, maybe offering specific, optimized 'cut-off' points via targeted training could be a useful tradeoff?

Fair enough, the "every other layer" is what we used for https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2 (the base model was originally 12 layers instead of 6), although we didn't use the adaptive layer loss.
This "every other layer" or "specific list" is also what we use in Model Distillation > Layer Reduction: https://sbert.net/examples/sentence_transformer/training/distillation/README.html

---

> weighting the similarity loss based on layer_idx (and not weighting the divergence loss if I recall correctly, it's been a while) seemed to me a bit arbitrary, but probably I missed something.

Makes sense. I believe this was based on the [2DMSE paper](https://arxiv.org/pdf/2402.14776), but they later rebranded and updated their work, so I can't really find it back :/ 
I can find their now work, though: https://github.com/SeanLee97/AnglE/blob/d29724e54ddf9dc55194e68fc8003d1d7a19209d/angle_emb/angle.py#L938-L996
But it doesn't really seem to match, e.g. they seem to divide by `1. + math.log(1 + layer_idx)`.

- Tom Aarsen
